other/huffmancoding/huffmancoding.go

```diff
@@ -0,0 +1,123 @@
+// details:
+// We implement the linear-time 2-queue method described here https://en.wikipedia.org/wiki/Huffman_coding.
+// It assumes that the list of symbol-frequencies is sorted.
+// author(s) [pedromsrocha](https://github.com/pedromsrocha)
+// see also huffmancoding_test.go
+
+// Package huffman provides primitives for doing Huffman compression, encodding and decoding.
+package huffman
+
+// A Node of an Huffman tree, which can either be a leaf or an internal node.
+// Each node has a weight.
+// A leaf node has an associated symbol, but no children (i.e., left == right == nil).
+// A parent node has a left and right child and no symbol (i.e., symbol == -1).
+type Node struct {
+	left   *Node
+	right  *Node
+	symbol rune
+	weight int
+}
+
+// A SymbolFreq is a pair of a symbol and its associated frequency.
+type SymbolFreq struct {
+	symbol rune
+	freq   int
+}
+
+// The algorithm operates on []SymbolFreq sorted by frequency
+type ByFreq []SymbolFreq
+
+func (x ByFreq) Len() int           { return len(x) }
+func (x ByFreq) Less(i, j int) bool { return x[i].freq < x[j].freq }
+func (x ByFreq) Swap(i, j int)      { x[i], x[j] = x[j], x[i] }
+
+// BuildTree returns the root Node of the Huffman tree by compressing listfreq.
+// The compression produces the most optimal code lengths, provided listfreq is ordered,
+// i.e.: listfreq[i] <= listfreq[j], whenever i < j.
+func BuildTree(listfreq ByFreq) Node {
+	if len(listfreq) < 1 {
+		panic("huffman: calling BuildTree with an empty list of symbol-frequency pairs")
+	}
+	q1 := make([]Node, len(listfreq))
+	q2 := make([]Node, 0, len(listfreq))
+	for i, x := range listfreq { // after the loop, q1 is a slice of leaf nodes representing listfreq
+		q1[i] = Node{left: nil, right: nil, symbol: x.symbol, weight: x.freq}
+	}
+	//loop invariant: q1, q2 are ordered by increasing weights
+	for len(q1)+len(q2) > 1 {
+		var node1, node2 Node
+		node1, q1, q2 = least(q1, q2)
+		node2, q1, q2 = least(q1, q2)
+		node := Node{left: &node1, right: &node2,
+			symbol: -1, weight: node1.weight + node2.weight}
+		q2 = append(q2, node)
+	}
+	if len(q1) == 1 { // returns the remaining node in q1, q2
+		return q1[0]
+	} else {
+		return q2[0]
+	}
+}
+
+// least removes the node with lowest weight from q1, q2.
+// It returns the node with lowest weight and the slices q1, q2 after the update.
+func least(q1 []Node, q2 []Node) (Node, []Node, []Node) {
+	if len(q1) == 0 {
+		return q2[0], q1, q2[1:]
+	}
+	if len(q2) == 0 {
+		return q1[0], q1[1:], q2
+	}
+	if q1[0].weight <= q2[0].weight {
+		return q1[0], q1[1:], q2
+	} else {
+		return q2[0], q1, q2[1:]
+	}
+}
+
+// BuildDict recursively traverses the Huffman tree pointed by node to build dict, that maps symbols to binary codes,
+// each code being prefixed by prefix.
+// left and right children are labelled with the booleans false and true, respectively.
+func BuildDict(node *Node, prefix []bool, dict map[rune][]bool) {
```

Comment:
Dict is a term that refers to a concept in `Python`, this is a Go repository. Try to name things using the Go conventions :pray: :smile:

Can you think of the actual name of this step of the code? Even the python [implementation](https://github.com/TheAlgorithms/Python/blob/67676c3b790d9631ea99c89f71dc2bf65e9aa2ca/compression/huffman.py#L69) is not using this particular step (from what I can tell - I skimmed it so maybe I missed something :smile:).
PS: dict is called a map in Go.

Reply:
I've opted to use the term `dict` in accordance with the terminology used in the [wiki page](https://en.wikipedia.org/wiki/Huffman_coding) of the Huffman coding algorithm, on which this implementation is based. However, I understand that this might violate Go conventions, so commit 7193e6e renames method `BuldDitct` to - I hope - a more descriptive name `GetEncoding`. This corresponds to the step of the algorithm in which the Huffman tree is traversed (a kind of tree flattening) to obtain the encoding map.